File: typescript/agentkit/src/action-providers/taskmarket/taskmarketActionProvider.ts
import { execFile } from "child_process";
import { z } from "zod";
import { ActionProvider } from "../actionProvider";
import { CreateAction } from "../actionDecorator";
import {
  ListTasksSchema,
  GetTaskSchema,
  GetMySubmissionsSchema,
  SubmitWorkSchema,
  CreateTaskSchema,
} from "./schemas";

/**
 * TaskmarketActionProvider is an action provider for Taskmarket
 * (https://taskmarket.dev) — an on-chain bounty marketplace for AI agents.
 *
 * It enables an agent to discover open bounty work, inspect a precise task,
 * and track its own submissions by delegating to the `taskmarket` CLI.
 * State-changing actions (`submit_work`, `create_task`) are gated behind an
 * explicit `confirm` input so funds and on-chain state are never touched
 * without user authorization.
 */
export class TaskmarketActionProvider extends ActionProvider {
  private readonly cliPath: string;

  private readonly timeoutMs: number;

  /**
   * Constructor for the TaskmarketActionProvider class.
   *
   * @param cliPath - Path to the taskmarket CLI executable (default "taskmarket")
   * @param timeoutMs - Timeout for CLI invocations in milliseconds
   */
  constructor(cliPath = "taskmarket", timeoutMs = 60_000) {
    super("taskmarket", []);
    this.cliPath = cliPath;
    this.timeoutMs = timeoutMs;
  }

  /**
   * Runs the taskmarket CLI with the given arguments and resolves with stdout.
   *
   * @param args - CLI arguments to pass through
   * @returns A promise resolving to the CLI stdout
   */
  private run(args: string[]): Promise<string> {
    return new Promise((resolve, reject) => {
      execFile(this.cliPath, args, { timeout: this.timeoutMs }, (error, stdout, stderr) => {
        if (error) {
          const detail = (stderr && stderr.trim()) || error.message;
          reject(new Error(detail));
          return;
        }
        resolve(stdout);
      });
    });
  }

  /**
   * Lists currently open Taskmarket tasks.
   *
   * @param args - Optional filters: status, limit, minimum reward, and tags
   * @returns A JSON string containing matching tasks or an error message
   */
  @CreateAction({
    name: "list_tasks",
    description: `This tool lists open tasks on Taskmarket, an on-chain bounty marketplace where agents earn USDC for completed digital work.

It takes the following optional inputs:
- status: task status filter (default "open")
- limit: maximum number of results (default 20)
- rewardMin: minimum reward in USDC
- tags: comma-separated tag filter

Important notes:
- Each result includes the task id (0x-prefixed hex), description, reward, expiry, and submission count
- Use get_task with the task id for full details before doing any work`,
    schema: ListTasksSchema,
  })
  async listTasks(args: z.infer<typeof ListTasksSchema>): Promise<string> {
    try {
      const cliArgs = ["task", "list", "--status", args.status, "--limit", String(args.limit)];
      if (args.rewardMin !== undefined && args.rewardMin !== null) {
        cliArgs.push("--reward-min", String(args.rewardMin));
      }
      if (args.tags) {
        cliArgs.push("--tags", args.tags);
      }
      return await this.run(cliArgs);
    } catch (error: unknown) {
      return `Error listing Taskmarket tasks: ${error instanceof Error ? error.message : String(error)}`;
    }
  }

  /**
   * Gets the full details of a single Taskmarket task.
   *
   * @param args - The task id to inspect
   * @returns A JSON string containing the task record or an error message
   */
  @CreateAction({
    name: "get_task",
    description: `This tool fetches the full details of one Taskmarket task by its 0x-prefixed hex id.

It takes the following inputs:
- taskId: the 0x-prefixed task id from list_tasks

Important notes:
- Returns description, reward, expiry, escrow transaction, submission count, awards, and lifecycle phase
- Check that submissionWindowOpen is true, stakeRequired is false, and the reward is escrowed before starting work`,
    schema: GetTaskSchema,
  })
  async getTask(args: z.infer<typeof GetTaskSchema>): Promise<string> {
    try {
      return await this.run(["task", "get", args.taskId]);
    } catch (error: unknown) {
      return `Error getting Taskmarket task: ${error instanceof Error ? error.message : String(error)}`;
    }
  }

  /**
   * Lists submissions made by the configured Taskmarket wallet.
   *
   * @returns A JSON string containing the wallet's submissions or an error message
   */
  @CreateAction({
    name: "my_submissions",
    description: `This tool lists all Taskmarket submissions made by the configured wallet, across all tasks.

Important notes:
- Requires the taskmarket CLI to be authenticated (taskmarket init)
- Use it to track whether submissions were accepted, rejected, or awarded`,
    schema: GetMySubmissionsSchema,
  })
  async mySubmissions(): Promise<string> {
    try {
      return await this.run(["task", "my-submissions"]);
    } catch (error: unknown) {
      return `Error listing Taskmarket submissions: ${error instanceof Error ? error.message : String(error)}`;
    }
  }

  /**
   * Submits a file as work for a Taskmarket task. Requires explicit confirmation.
   *
   * @param args - Task id, file path, artifact role, and confirmation flag
   * @returns The CLI submission result or an error message
   */
  @CreateAction({
    name: "submit_work",
    description: `This tool submits a completed work file to a Taskmarket task.

It takes the following inputs:
- taskId: the 0x-prefixed task id
- filePath: local path of the deliverable file
- role: artifact role (preview, source, final, attachment; default final)
- confirm: must be explicitly set to true by the user to submit

Important notes:
- Submission anchors the artifact on-chain and cannot be undone; never call this without explicit user approval
- Re-check the task with get_task immediately before submitting to confirm the window is still open
- Returns the submission id when successful`,
    schema: SubmitWorkSchema,
  })
  async submitWork(args: z.infer<typeof SubmitWorkSchema>): Promise<string> {
    if (!args.confirm) {
      return "Submission not confirmed. Set confirm: true to submit work to this task. Submitting is irreversible and anchors the artifact on-chain.";
    }
    try {
      return await this.run([
        "task",
        "submit",
        args.taskId,
        "--file",
        args.filePath,
        "--role",
        args.role,
      ]);
    } catch (error: unknown) {
      return `Error submitting Taskmarket work: ${error instanceof Error ? error.message : String(error)}`;
    }
  }

  /**
   * Creates a new Taskmarket task, escrowing the reward in USDC. Requires explicit confirmation.
   *
   * @param args - Description, reward, duration, tags, and confirmation flag
   * @returns The CLI creation result or an error message
   */
  @CreateAction({
    name: "create_task",
    description: `This tool creates a new Taskmarket task so external workers can complete it for a USDC reward.

It takes the following inputs:
- description: full task specification shown to workers
- rewardUsdc: reward in USDC — this amount is escrowed from the wallet
- durationHours: hours until the task expires
- tags: optional comma-separated tags
- confirm: must be explicitly set to true by the user to create and escrow

Important notes:
- Creating a task spends the reward amount in USDC; never call this without explicit user approval
- Only create tasks for work the user actually wants delegated`,
    schema: CreateTaskSchema,
  })
  async createTask(args: z.infer<typeof CreateTaskSchema>): Promise<string> {
    if (!args.confirm) {
      return `Task creation not confirmed. Set confirm: true to escrow ${args.rewardUsdc} USDC and publish this task.`;
    }
    try {
      const cliArgs = [
        "task",
        "create",
        "--description",
        args.description,
        "--reward",
        String(args.rewardUsdc),
        "--duration",
        String(args.durationHours),
      ];
      if (args.tags) {
        cliArgs.push("--tags", args.tags);
      }
      return await this.run(cliArgs);
    } catch (error: unknown) {
      return `Error creating Taskmarket task: ${error instanceof Error ? error.message : String(error)}`;
    }
  }

  /**
   * Checks if the Taskmarket action provider supports the given network.
   * Taskmarket settles on Base but discovery is network-agnostic, so this always returns true.
   *
   * @returns True, as Taskmarket actions are supported on all networks.
   */
  supportsNetwork(): boolean {
    return true;
  }
}

/**
 * Creates a new instance of the TaskmarketActionProvider.
 *
 * @returns A new TaskmarketActionProvider instance
 */
export const taskmarketActionProvider = () => new TaskmarketActionProvider();

File: python/coinbase-agentkit/coinbase_agentkit/action_providers/taskmarket/taskmarket_action_provider.py
"""TaskMarket action provider for delegated work discovery and drafting.

The provider is intentionally read-only/safe by default: it can browse public TaskMarket work
and prepare an auditable task draft, but it does not create, fund, or accept tasks without an
explicit host-application approval flow.
"""

from __future__ import annotations

import json
from typing import Any

import requests

from ...network import Network
from ...wallet_providers import WalletProvider
from ..action_decorator import create_action
from ..action_provider import ActionProvider
from .schemas import BrowseTaskMarketTasksSchema, TaskMarketTaskDraftSchema

TASKMARKET_API_URL = "https://api.taskmarket.dev"


class TaskMarketActionProvider(ActionProvider[WalletProvider]):
    """Provides safe TaskMarket discovery and delegation-draft actions."""

    def __init__(self, api_url: str = TASKMARKET_API_URL):
        """Initialize the TaskMarket provider.

        Args:
            api_url: Base TaskMarket API URL. Override in tests or self-hosted deployments.

        """
        super().__init__("taskmarket", [])
        self.api_url = api_url.rstrip("/")

    @create_action(
        name="browse_taskmarket_tasks",
        description="""Browse public TaskMarket tasks as a delegation option for work that is better handled by external workers. This action is read-only and never spends funds. Use filters such as tag and max_submissions to find relevant, lower-competition tasks.""",
        schema=BrowseTaskMarketTasksSchema,
    )
    def browse_taskmarket_tasks(self, args: dict[str, Any]) -> str:
        """Browse public TaskMarket tasks.

        Args:
            args: limit, optional tag, and optional max_submissions filters.

        Returns:
            JSON string with simplified open-task data or an error payload.

        """
        validated_args = BrowseTaskMarketTasksSchema(**args)
        try:
            response = requests.get(
                f"{self.api_url}/api/tasks",
                params={"status": "open", "limit": validated_args.limit},
                timeout=20,
            )
            if not response.ok:
                return json.dumps(
                    {"success": False, "error": f"HTTP error! status: {response.status_code}"}
                )

            payload = response.json()
            raw_tasks = payload.get("tasks") or payload.get("data", {}).get("tasks") or []
            tasks = []
            for task in raw_tasks:
                tags = task.get("tags") or []
                submission_count = task.get("submissionCount") or task.get("submission_count") or 0
                if validated_args.tag and validated_args.tag.lower() not in [
                    str(tag).lower() for tag in tags
                ]:
                    continue
                if (
                    validated_args.max_submissions is not None
                    and submission_count > validated_args.max_submissions
                ):
                    continue
                task_id = task.get("id")
                tasks.append(
                    {
                        "id": task_id,
                        "url": f"https://taskmarket.dev/task/{task_id}" if task_id else None,
                        "title": task.get("title") or (task.get("description") or "")[:120],
                        "mode": task.get("mode"),
                        "status": task.get("status"),
                        "netReward": task.get("netReward")
                        or task.get("net_reward")
                        or task.get("netRewardBaseUnits")
                        or task.get("rewardBaseUnits"),
                        "submissionCount": submission_count,
                        "awardCount": task.get("awardCount") or task.get("award_count") or 0,
                        "tags": tags,
                        "submissionWindowOpen": task.get("submissionWindowOpen"),
                    }
                )

            return json.dumps(
                {
                    "success": True,
                    "tasks": tasks,
                    "returned": len(tasks),
                    "safety": "read_only_no_spend",
                }
            )
        except Exception as exc:
            return json.dumps({"success": False, "error": str(exc)})

    @create_action(
        name="prepare_taskmarket_task_draft",
        description="""Prepare a TaskMarket delegation draft with budget, deadline, deliverable, and acceptance criteria. This action does not create or fund a task; the host app must display the draft and get explicit user approval before any wallet/payment action.""",
        schema=TaskMarketTaskDraftSchema,
    )
    def prepare_taskmarket_task_draft(self, args: dict[str, Any]) -> str:
        """Prepare an auditable TaskMarket task draft without spending funds."""
        draft = TaskMarketTaskDraftSchema(**args)
        return json.dumps(
            {
                "success": True,
                "draft": draft.model_dump(),
                "nextStep": "Present this draft to the user for explicit approval before creating or funding a TaskMarket task.",
                "safety": {
                    "spendsFunds": False,
                    "createsTask": False,
                    "requiresExplicitApproval": True,
                },
            }
        )

    def supports_network(self, network: Network) -> bool:
        """TaskMarket discovery/drafting is network agnostic and does not sign transactions."""
        return True


def taskmarket_action_provider(api_url: str = TASKMARKET_API_URL) -> TaskMarketActionProvider:
    """Create a TaskMarket action provider."""
    return TaskMarketActionProvider(api_url=api_url)
